### PR TITLE
IMAP Hi Histogram L1A processing

### PR DESCRIPTION
**Pull request by tmplummer**
# Change Summary
Add L1A processing of Hi Hist packets.

## Overview
I am opening this PR before finalizing test coverage in order to get feedback before fully fleshing out test coverage.

## New Dependencies
None

## New Files
- imap_processing/hi/l1a/histogram.py
   - House the L1A processing specific to the hist packet (SCI_CNT)

## Deleted Files
None

## Updated Files
- imap_processing/hi/hi_cdf_attrs.py
   - add histogram specific cdf attributes
- imap_processing/hi/l1a/hi_l1a.py
   - add code into branch for processing histogram packets
   - add branch for unexpected APIDs with logging
- imap_processing/hi/packet_definitions/hi_packet_definition.xml
   - Remove comment about Acquisition Start Time in description of ESA_STEP. This was an artifact from IBEX.
- imap_processing/test/hi/test_l1a.py
   - preliminary test coverage for L1A processing of histogram packets

## Testing
One unit test added - I plan to refine this test and possibly add more after getting initial feedback

### CODE REVIEW

File: imap_processing/hi/hi_cdf_attrs.py [hunk 1]
```diff
@@ -158,3 +159,54 @@
     var_type="support_data",
     variable_purpose="PRIMARY",
 )
+
+# Histogram attributes
+hi_hist_l1a_global_attrs = GlobalDataLevelAttrs(
+    data_type="L1A_CNT>Level-1A Histogram",
+    logical_source="imap_hi_l1a_hist",
+    logical_source_desc=("IMAP-HI Instrument Level-1A Histogram Data."),
```
Comment (greglucas) [inline imap_processing/hi/hi_cdf_attrs.py]: I think we've been using Hi and Lo rather than all caps for instrument names, but someone else can probably confirm.
```suggestion
    logical_source_desc=("IMAP-Hi Instrument Level-1A Histogram Data."),
```

File: imap_processing/hi/hi_cdf_attrs.py [hunk 2]
```diff
@@ -158,3 +159,54 @@
...
+    instrument_base=hi_base,
+)
+
+hi_hist_l1a_angle_attrs = AttrBase(
+    validmin=GlobalConstants.INT_FILLVAL,
+    validmax=GlobalConstants.INT_MAXVAL,
```
Comment (greglucas) [inline imap_processing/hi/hi_cdf_attrs.py]: Are these supposed to be angles in degrees?
```suggestion
    validmin=0,
    validmax=360,
```

File: imap_processing/hi/hi_cdf_attrs.py [hunk 3]
```diff
@@ -158,3 +159,54 @@
...
+    format="I9",
+    label_axis="ANGLE",
+    display_type="time_series",
+    catdesc=("Angle bin" "Bin centers for Histogram data."),
```
Comment (greglucas) [inline imap_processing/hi/hi_cdf_attrs.py]:
```suggestion
    catdesc=("Angle of bin centers for Histogram data."),
```

File: imap_processing/hi/hi_cdf_attrs.py [hunk 4]
```diff
@@ -158,3 +159,54 @@
...
+    fill_val=GlobalConstants.INT_FILLVAL,
+    var_type="support_data",
+)
+
+hi_hist_l1a_esa_step_attrs = ScienceAttrs(
+    validmin=0,
+    validmax=2**4 - 1,
+    depend_0="epoch",
+    format="I2",
+    units="int",
```
Comment (greglucas) [inline imap_processing/hi/hi_cdf_attrs.py]: Are these unitless quantities? I think that is supposed to be left empty in the units category, not the datatype.
```suggestion
    units="",
```

File: imap_processing/hi/l1a/hi_l1a.py [hunk 5]
```diff
@@ -26,27 +27,31 @@ def hi_l1a(packet_file_path: str):
...
         elif apid == HIAPID.H45_APP_NHK:
             logger.info(
                 "Processing housekeeping data for [%s] packets", HIAPID.H45_APP_NHK.name
             )
-            data = create_dataset(grouped_data[apid])
-            processed_data.append(data)
+            data = hk_create_dataset(grouped_data[apid])
+        else:
+            logger.warning(f"Encountered unexpected APID [{apid}]")
+            continue
```
Comment (greglucas) [inline imap_processing/hi/l1a/hi_l1a.py]: Is this OK to have happen, or should we be raising RunTimeError when this occurs to attempt to abort?

File: imap_processing/hi/l1a/histogram.py [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+    coords["angle"] = xr.DataArray(
+        np.arange(2, 360, 4),
+        name="angle",
+        dims=["angle"],
+        attrs=hi_cdf_attrs.hi_hist_l1a_angle_attrs.output(),
+    )
+
+    data_vars = dict()
+    data_vars["esa_step"] = xr.DataArray(
+        np.empty(len(packets), np.dtype("uint8")),
```
Comment (greglucas) [inline imap_processing/hi/l1a/histogram.py]: minor nit: I have a slight preference for being explicit with adding a keyword argument to the empty method call and using the builtin `np.uint8` for consistency with other places we've been using dtypes in the codebase.
```suggestion
        np.empty(len(packets), dtype=np.uint8),
```

File: imap_processing/hi/l1a/histogram.py [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+            fieldname="ACQ_START",
+            label_axis="ACQ_START",
+        ).output(),
+    )
+
+    # This dictionary defines the names and order of the 24, 90 element counter arrays
+    # TODO: This implementation will result in a CDF with variables consisting of the
+    #    counter_def_dict key + tuple string. Need to verify if this is OK with
+    #    instrument team
+    counter_def_dict = {
```
Comment (greglucas) [inline imap_processing/hi/l1a/histogram.py]: I found it hard to parse the nested for loops at first glance, so I am just giving a quick suggestion in case this appeals to you or sparks another idea. But, the way it currently works is fine as well if you like that style better.
```python
qual_list = ["QUAL_" + x for x in ("AB", "C1C2", "AC1", "BC1", "ABC1", "AC1C2", "BC1C2", "ABC1C2")]
long_list = ["LONG_" + ...]
misc_list = ["MISC_" + ...]
for key_counter in qual_list + long_list + misc_list:
    data_vars[key_counter] = ...
```

File: imap_processing/hi/l1a/histogram.py [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+
+    # unpack the packets data into the Dataset
+    for i_epoch, packet in enumerate(packets):
+        dataset.epoch.data[i_epoch] = calc_start_time(
+            packet.data["CCSDS_MET"].raw_value
+        )
+        dataset.esa_step.data[i_epoch] = packet.data["ESA_STEP"].raw_value
+        dataset.acq_start.data[i_epoch] = packet.data["ACQ_START"].raw_value
+
+        counters = ConstBitStream(bin=packet.data["COUNTERS"].raw_value)
```
Comment (greglucas) [inline imap_processing/hi/l1a/histogram.py]: I've been trying to limit our usage of bitstring when possible for performance reasons. Are we able to convert these values ourselves with native Python functionality? Here is a quick previous example: https://github.com/IMAP-Science-Operations-Center/imap_processing/pull/309

File: imap_processing/hi/packet_definitions/hi_packet_definition.xml [hunk 9]
```diff
@@ -123,12 +123,13 @@
 			</xtce:Parameter>
             <!-- SCI_DE end -->
             <!-- SCI_CNT start -->
-            <xtce:Parameter name="ESA_STEP" parameterTypeRef="UINT16" shortDescription="Lower 4-bits represent the ESA Step Number: 4-bit ESA Step Enumeration 12-bit Start of Acquisition Time">
-				<xtce:LongDescription>
-					Lower 4-bits represent the ESA Step Number:
-					4-bit ESA Step Enumeration
-					12-bit Start of Acquisition Time
-				</xtce:LongDescription>
+            <xtce:Parameter name="ACQ_START" parameterTypeRef="UINT12">
```
Comment (greglucas) [inline imap_processing/hi/packet_definitions/hi_packet_definition.xml]: Nice :)

File: imap_processing/hi/hi_cdf_attrs.py [hunk 10]
```diff
@@ -158,3 +159,54 @@
...
+    data_type="L1A_CNT>Level-1A Histogram",
+    logical_source="imap_hi_l1a_hist",
+    logical_source_desc=("IMAP-HI Instrument Level-1A Histogram Data."),
+    instrument_base=hi_base,
+)
+
+hi_hist_l1a_angle_attrs = AttrBase(
+    validmin=GlobalConstants.INT_FILLVAL,
+    validmax=GlobalConstants.INT_MAXVAL,
+    format="I9",
```
Comment (tech3371) [inline imap_processing/hi/hi_cdf_attrs.py]: If we make @greglucas 's suggestion on min and max, then we can update this format too to match that.
```suggestion
    format="I3",
```

File: imap_processing/hi/l1a/hi_l1a.py (hunk 5 above)
Comment (tech3371) [reply imap_processing/hi/l1a/hi_l1a.py]: Good question. I think we should log it because if we have other good apid that we could process, it would be good to process those. But good point about how to raise this error without aborting all.

File: imap_processing/hi/hi_cdf_attrs.py (hunk 1 above)
Comment (tmplummer) [reply imap_processing/hi/hi_cdf_attrs.py]: I confirmed that Lo cdf global attributes are not all caps. I will change the other occurrences of "HI" in this file.

File: imap_processing/hi/hi_cdf_attrs.py (hunk 4 above)
Comment (tmplummer) [reply imap_processing/hi/hi_cdf_attrs.py]: Right. I don't know what the units are for this field. I would think that it is some fraction of a second, but am unsure.
@tech3371 With Vivek leaving/off the project who should I ask for clarification about what this value is?

File: imap_processing/hi/l1a/hi_l1a.py (hunk 5 above)
Comment (tmplummer) [reply imap_processing/hi/l1a/hi_l1a.py]: Valid point.

File: imap_processing/hi/hi_cdf_attrs.py (hunk 4 above)
Comment (tech3371) [reply imap_processing/hi/hi_cdf_attrs.py]: Yaxi, yaxi.liu@swri.org

File: imap_processing/hi/l1a/histogram.py [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+        dims=["epoch"],
+        attrs=dataclasses.replace(
+            hi_cdf_attrs.hi_hist_l1a_acq_start_attrs,
+            catdesc=packets[0].data["ACQ_START"].long_description,
+            fieldname="ACQ_START",
+            label_axis="ACQ_START",
+        ).output(),
+    )
+
+    # This dictionary defines the names and order of the 24, 90 element counter arrays
```
Comment (tech3371) [inline imap_processing/hi/l1a/histogram.py]: This is what I was thinking of doing if it helps. https://github.com/tech3371/imap_processing/blob/hi_l1a/imap_processing/hi/l1a/science_histogram.py

Then You could have counters dimension with those values, `"AB", "C1C2", "AC1", "BC1",...`, angle dimension like you had.

File: imap_processing/hi/l1a/histogram.py (hunk 8 above)
Comment (tech3371) [reply imap_processing/hi/l1a/histogram.py]: I remember this discussion. But for this case, it seems like bitstring may be the best choice since we need to chunk data into `uint12`. Other ways was not readable or too much workaround.

Comment (greglucas) [reply imap_processing/hi/l1a/histogram.py]: Would this work to get a list of ints from that data? This doesn't seem too bad to me.
```python
data_string = packet.data["COUNTERS"].raw_value
list_of_ints = [int(data_string[i*12:(i+1)*12], 2) for i in range(90*24)]
```

Profiling it is 10x faster:
```
❯ python -m timeit 'import bitstring; x = "10" * 6 * 90 * 24; counters = bitstring.ConstBitStream(bin=x); data = counters.readlist("2160*uint12")'
50 loops, best of 5: 7.66 msec per loop
❯ python -m timeit 'import bitstring; x = "10" * 6 * 90 * 24; counters = bitstring.ConstBitStream(bin=x); data = [int(x[i*12:(i+1)*12], 2) for i in range(90*24)]'
500 loops, best of 5: 742 usec per loop
```

Comment (tmplummer) [reply imap_processing/hi/l1a/histogram.py]: Wow, 10x faster!
I have essentially the same implementation staged to commit. I am working through a few other changes in this PR as well.

Comment (tech3371) [reply imap_processing/hi/l1a/histogram.py]: I am amazed by solutions you guys come up with. Nice!

File: imap_processing/hi/hi_cdf_attrs.py (hunk 4 above)
Comment (tmplummer) [reply imap_processing/hi/hi_cdf_attrs.py]: I might have resolution on what this field is and what the units should be later today. Other units have been removed.

File: imap_processing/hi/l1a/hi_l1a.py (hunk 5 above)
Comment (tmplummer) [reply imap_processing/hi/l1a/hi_l1a.py]: I'm not sure how to resolve this. Getting an unexpected APID would indicate a bug in something upstream that didn't filter APIDs correctly or called the wrong processing function for a given L0 file. I don't think that we should plan for this to happen regularly so it seems like raising a RunTimeError would be appropriate in order to fail loudly.

Comment (greglucas) [reply imap_processing/hi/l1a/hi_l1a.py]: I share your concern about a logger warning not being loud enough and possibly going undetected which is why I've been stressing the raise instead of warn. But, I don't have a major preference on that and it is likely not a big deal. You can choose what you feel is appropriate here and we can change things later if we need to, I don't think anyone is blocking over this question.

File: imap_processing/hi/l1a/histogram.py [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+        # unpack 24 arrays of 90 12-bit unsigned integers
+        counters_binary_data = packet.data["COUNTERS"].raw_value
+        counter_ints = [
+            int(counters_binary_data[i * 12 : (i + 1) * 12], 2) for i in range(90 * 24)
+        ]
+        # populate the dataset with the unpacked integers
+        for i_counter, counter in enumerate(
+            (*QUALIFIED_COUNTERS, *LONG_COUNTERS, *TOTAL_COUNTERS)
+        ):
+            dataset[counter][i_epoch] = counter_ints[
```
Comment (greglucas) [inline imap_processing/hi/l1a/histogram.py:74]: Above you used `.data[i_epoch]`, but here you use just `[i_epoch]` is there a difference between these setters?

**Review by greglucas — APPROVED** [review]
Nice work!

**Comment by greglucas** [comment]
You do still need to appease pre-commit it looks like before merging.

File: imap_processing/hi/l1a/histogram.py (hunk 12 above)
Comment (tmplummer) [reply imap_processing/hi/l1a/histogram.py:74]: Good catch. I have no experience with xarray.DataArray. Both methods of setting values appear to do the same thing but I believe the correct/preferred way is without the `data` attribute. I'll make that change.

Comment (tmplummer) [reply imap_processing/hi/l1a/histogram.py:74]: Interestingly enough, if one tries to modify a coordinate variable without using the `data` attribute, xarray raises `TypeError: IndexVariable values cannot be modified`. Interestingly, using the `data` attribute has the desired effect with no error.

Comment (greglucas) [reply imap_processing/hi/l1a/histogram.py:74]: Ahh, so the issue is with updating the coordinates only then? And I suppose it is because you are looping through this only updating one at a time instead of wanting to use a full `assign()` or `assign_coords()` that would reset the entire array. With the `.data` you are accessing the underlying numpy array and then mutating that without touching any of the metadata held within the xarray class.

Good to know, thanks for digging in!

**Comment by bourque** [comment]
@all-contributors please add @subagonsouth for code!
